### ISSUE TITLE
Cannot find module '../compiled/src/app/app.module.ngfactory'

### ISSUE THREAD

**Issue by dherges** [AngularClass/angular-starter#1458]
**Note: for support questions, please use one of these channels:** [Chat: AngularClass.slack](http://angularclass.com/member-join/) or [Twitter: @AngularClass](https://twitter.com/AngularClass)

* **I'm submitting a ...**
[x] bug report
[x] feature request
[ ] question about the decisions made in the repository

* **Do you want to request a *feature* or report a *bug*?**

Both.

* **What is the current behavior?**

When ``compiled`` directory is empty or non-existing, running ``npm run build:dev`` results in an error from awesome-typescript-loader:

```
ERROR in [at-loader] src/main.browser.aot.ts:10:36 
    Cannot find module '../compiled/src/app/app.module.ngfactory'.
```

* **What is the expected behavior?**

Build should not produce errors even if ``compiled`` directory doesn't exist or is empty, since directory is not needed for a dev build without aot.

* **What is the motivation / use case for changing the behavior?**

Was able to work around this by:

in ``webpack.common.js``
```js
            {
              loader: 'awesome-typescript-loader',
              options: {
                configFileName: isProd ? 'tsconfig.webpack.json' : 'tsconfig.json'
              }
            },
```

in ``tsconfig.json``:

```json
  "exclude": [
    "node_modules",
    "dist",
    "src/main.browser.aot.ts"
  ],

```

Then ``npm run build:dev`` finishes without errors. Still ``npm run build:prod`` causes the same error message.

* **Please tell us about your environment:**

- Angular version: 2.4.5
- Browser: webpack 2.2.0

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)

**Comment by mbenzenhoefer** [comment]
+1
I am currently having the same issue with this setup

**Comment by shlomiassaf** [comment]
This error is fired by the `CheckerPlugin` from `awesome-typescript-loader`

The checker plugin process all typescript files in the `src` directory while the `awesome-typescript-loader` process files that webpack provides starting from the root (entryPoint).

The end result is `src/main.browser.aot.ts` being process by the `CheckerPlugin` and if there is no content in the compile directory there is an error.

This is why we get a type error but not a bundling error, this is a "virtual" error that does not effect the end result, runtime will have no error.

The workaround is not a viable solution, the `tsconfig.json` is for the test environment and it will require syncing up both configs every time, if we need some changes to the test environment (and only test) we will face an issue.

There is a [PR proposal](https://github.com/AngularClass/angular2-webpack-starter/pull/1378) that tried to make typescript config dynamic but it makes things a bit more complicated for the end developer.

The best solution is typescript to accepts commonJS modules instead of json file :)

**Comment by dherges** [comment]
Ok, I see. Looks like we're going with yet another ``tsconfig.*.json`` file that is going to be used for non-AoT builds ... maybe ``tsconfig.dev.json`` or ``tsconfig.jit.json`` ... do you think this is a useful thing to have in this starter repo?

**Comment by shlomiassaf** [comment]
@dherges not sure if there will be another file, see s-panferov/awesome-typescript-loader#364

**Comment by dherges** [comment]
@shlomiassaf I don't get it. Right now, we have two files in this repo:

``tsconfig.json`` for tests
``tsconfig.webpack.json`` for webpack build

my suggestion is to add another one:

``tsconfig.jit.json`` for JiT compiler builds

plus, we could rename ``tsconfig.webpack.json`` to ``tsconfig.aot.json`` to make it clear, that this one is used for AoT compiler builds.

**Comment by shlomiassaf** [comment]
Did you look at s-panferov/awesome-typescript-loader#364

Instead of complicating the build with another static JSON configuration file that issue offers dynamic control over the scope of the checker.

Now, it might seem simple enough to add another `tsconfig.jit.json` but the reality is that someone has to maintain it...

Every change made in one `tsconfig.XXX.json`has to be replicated into all other tsconfig files.

We already have hard time dealing with the different TEST environment setup and once in a while we discover that a feature that was added to one config was not added to another....

Not to mention we have multiple webpack configs and multiple tsconfig configs.

The goal is to consolidate the configuration and not spread it, so I don't think adding a new file is the right approach.

Again, s-panferov/awesome-typescript-loader#364 offers a good solution, a little patience and we will fix this.

Of course, I'm not calling the shots, so please lets get input from @d3viant0ne @gdi2290 @colinskow

**Comment by dherges** [comment]
Thanks for your clarification. I understand your points. At the same time, I like to see the resulting configuration. Having "plain" json files in the repository allows to compare quickly compare different tsconfigs. I fully agree that this spreads out configuration and makes it harder to maintain. So, from my point of view, this is about maintainability vs. debuggability (does that word even exist?!?!): with dynamic config, I'd like to have some utility that prints the configuration. Maybe, I should reach out to s-panferov/awesome-typescript-loader then 💯

**Comment by shlomiassaf** [comment]
@dherges The tradeoff is clear, the balance is not.

The current status is that:
  - AOT compilation in angular is not that trivial, lot's of factors people don't realise.
  - Running unit tests with Karma requires a completely unique configuration
 
Add to that HMR, unique loaders, special production settings, linting etc.. 
The end result is a complicated process, to complicated if you ask me.

I help a lot of friends with setups, its a real blocker for most and this is why the angular-cli is a must, and I hope it will allow the freedom we get here, one day.

To sum up, factoring all of the above I think we should avoid spreading as much as possible.

**Comment by colinskow** [comment]
This is also fixed by #1378, but s-panferov/awesome-typescript-loader#364 will provide a better solution. We should wait to avoid over-complicating things.

And cloning the latest version of this repository on Mac and Windows I am able to do `npm run build:prod` just fine without any errors.
